lib/MusicBrainz/Server/Data/Release.pm

```diff
@@ -880,54 +883,155 @@ sub can_merge {
+        $target_count{$_->{id}}++ for @{$possible_merge->{old_recordings}};
     }
 
-    my %recording_by_position;
-    for my $m_pos (keys %medium_by_position) {
-        # must have at least two mediums
-        my @mediums = @{ $medium_by_position{$m_pos} };
-        next if @mediums <= 1;
-        # all mediums must have the same number of tracks
-        my $track_count = $mediums[0]->track_count;
-        next if grep { $_->track_count != $track_count } @mediums;
-        # group recordings by track position
-        $recording_by_position{$m_pos} = {};
-        for my $medium (@mediums) {
-            for my $tr ($medium->all_tracks) {
-                my $tr_pos = $tr->position;
-                if (exists $recording_by_position{$m_pos}->{$tr_pos}) {
-                    push @{ $recording_by_position{$m_pos}->{$tr_pos} }, $tr->recording;
-                }
-                else {
-                    $recording_by_position{$m_pos}->{$tr_pos} = [ $tr->recording ];
-                }
+    # MBS-8614. Track recording merges, to resolve cases where a recording is
+    # a merge source on one track (after which it gets deleted), and a merge
+    # target on another track (in which case we should instead use the ID of
+    # the target from the first merge). Example where recording 3 should be
+    # merged into recording 2:
+    # 1 -> 2
+    # 3 -> 1
+    my %merge_targets;
+
+    $possible_merges = [grep {
+        # Assigned to avoid confusion with unrelated uses of $_ below.
+        my $possible_merge = $_;
+
+        my ($new_recording, $old_recordings) = @{$possible_merge}{qw(
+            new_recording
+            old_recordings
+        )};
+
+        # As mentioned above, remove old (source) recordings that have
+        # multiple targets.
+        #
+        # Perl fact: this assignment modifies the array in $possible_merge.
+        @{$old_recordings} = grep { $target_count{$_->{id}} == 1 } @{$old_recordings};
+
+        $merge_targets{$_->{id}} = $new_recording for @{$old_recordings};
+
+        # Make sure there's at least one merge source left for this target.
+        # (This is the expression for which `grep` is evaluated on.)
+        scalar @{$old_recordings} > 0
+    } @{$possible_merges}];
+
+    my %recording_merges;
+    for my $possible_merge (@{$possible_merges}) {
+        my ($new_recording, $old_recordings) = @{$possible_merge}{qw(
+            new_recording
+            old_recordings
+        )};
+
+        $new_recording = $merge_targets{$new_recording->{id}} // $new_recording;
+        my $new_id = $new_recording->{id};
+
+        for my $old_recording (@{$old_recordings}) {
+            my $old_id = $old_recording->{id};
+
+            # If two recordings' positions are swapped (e.g. recording 1 is being
+            # merged into recording 2, and recording 2 is being merged into
+            # recording 1), then we don't merge them in that case, because it's
+            # probably not intentional.
```

Comment:
I'd expect this to also cancel the merge.

Reply:
(I'd also suggest to not even allow entering a merge edit if this is the case when the edit is entered)

Reply:
That's handled by commit "Reject recording merge cycles"